File: Haskell/14-maximal-distance.hs
import Control.Monad (replicateM)

data Point = Point { xcoord :: Int, index :: Int }


readPoints :: Int -> IO [Point]
readPoints n = do
    xs <- replicateM n readLn
    return $ map (\(x, ix) -> Point x ix) $ zip xs [1..]


maxDistances :: [Point] -> [(Int, Int)]
maxDistances points = let first @ (Point _ ix) = head points in go (tail points) [(ix, ix)] first first
    where
        go :: [Point] -> [(Int, Int)] -> Point -> Point -> [(Int, Int)]
        go [] acc _ _ = reverse acc
        go (p : xss) acc left right =
            let left' = if xcoord p < xcoord left then p else left
                right' = if xcoord p > xcoord right then p else right
            in go xss ((index left', index right') : acc) left' right'


main :: IO()
main = do
    n <- readLn
    xs <- readPoints n
    let result = maxDistances xs
    mapM_ (\(i, j) -> putStrLn $ unwords $ map show [i, j]) result

File: Python/14_maximal_distance.py
import sys

from collections import namedtuple

Point = namedtuple('Point', ['x', 'ix'])


def read_points(reader, n):
    points = []
    for ix in range(n):
        x = int(next(reader))
        points.append(Point(x, ix + 1))
    return points


def max_distances(points):
    first_point = points[0]
    left = first_point
    right = first_point
    result = [(left.ix, right.ix)]
    for p in points[1:]:
        if p.x < left.x:
            left = p
        if p.x > right.x:
            right = p
        result.append((left.ix, right.ix))
    return result


def main():
    reader = sys.stdin
    n = int(next(reader))
    xs = read_points(reader, n)
    result = max_distances(xs)
    for points in result:
        print(' '.join([str(x) for x in points]))


if __name__ == '__main__':
    main()
